Fix relative imports in src code, relocate src

jakemcdermott opened the pull request:
(Lots of files modified but the changes are low-risk and automated)

This PR is a follow-up to https://github.com/ansible/awx/pull/6954, specifically the part where we [removed webpack aliases](https://github.com/ansible/awx/pull/6954/commits/0f335170c537d656b92d25bf4a8897634555596d) to pick up `react-scripts`.

##### SUMMARY
- [x] Add the [jsconfig.json](https://create-react-app.dev/docs/importing-a-component/#absolute-imports) w/  `baseUrl` to enable absolute imports from `src`. Replaces all the `../../../../../` importing
- [x]  `jsx` -> `js` extension renaming 
- [x] hooks directory
- ~[ ] ui_next -> ui~ see: https://github.com/ansible/awx/issues/10676

For posterity, the script used to fix the relative imports is included in the commit history.

```
python3 tools/fixrelative.py awx/ui_next/src
npm --prefix=awx/ui_next run lint -- --fix
npm --prefix=awx/ui_next run prettier
```

```python3
import argparse
import glob
import os
from shutil import move, copymode
from tempfile import mkstemp


def get_new_import_string(old_import_str, root):
    if not root.startswith("./"):
        raise Exception("root must start with './'")
    name = root.replace("./", "")
    left, right = old_import_str.split("from")
    left += "from '"
    _, trailing = right.split(name)

    return left + name + trailing


roots = [
    "./api",
    "./components",
    # "./constants",
    "./contexts",
    "./screens",
    "./types",
    "./util",
]


def get_root(line):
    matched_root = None
    for root in roots:
        if root in line:
            matched_root = root
            break
    if "jest" in line:
        matched_root = None
    return matched_root


def find_and_replace_roots(file_path, root_dir, preview):
    fh, temp_path = mkstemp()
    has_logged_file_name = False
    with os.fdopen(fh, "w") as new_file:
        with open(file_path) as old_file:
            for (line_number, line) in enumerate(old_file):
                matched_root = get_root(line)
                if matched_root:
                    new_line = get_new_import_string(line, matched_root)
                    if not preview:
                        new_file.write(new_line)
                    if not has_logged_file_name:
                        log_file_replacement(root_dir, file_path)
                        has_logged_file_name = True
                    log_line_replacement(line_number, line, new_line)
                elif not preview:
                    new_file.write(line)

    if not preview:
        copymode(file_path, temp_path)
        os.remove(file_path)
        move(temp_path, file_path)


def log_line_replacement(line_number, line, new_line):
    display_line = line.replace(os.linesep, "")
    display_new_line = new_line.replace(os.linesep, "")
    print(f"\t (line {line_number}): {display_line} --> {display_new_line}")


def log_file_replacement(root_dir, file_path):
    display_path = os.path.relpath(file_path, root_dir)
    print("")
    print(f"{display_path}:")


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("root_dir", help="Root directory")
    parser.add_argument("--preview", help="Preview (no write)", action="store_true")
    args = parser.parse_args()
    return args


def run():
    args = parse_args()
    search_path = args.root_dir + "**/**/*.js*"

    for file_path in glob.iglob(search_path, recursive=True):
        find_and_replace_roots(file_path, args.root_dir, args.preview)


if __name__ == "__main__":
    run()

```

jakemcdermott committed:
Enable absolute imports

jakemcdermott committed:
Add script to switch relative imports

AlexSCorey commented:
@jakemcdermott  Is this PR supposed to remove `.jsx` extension names?  Looks like there are a lot of `.jsx` files left.

jakemcdermott commented:
@AlexSCorey Yes, that's right - thanks. I'll add it to the script.

AlexSCorey commented:
triggered e2e tests

nixocio reviewed (approved):
I got this code locally, changes seem to be right. I was able to re-create the dev and production build without failures.  @jakemcdermott I noticed some references for `.jsx` on the `.po` files. Do we need to generate those again?

nixocio commented:
@jakemcdermott we could also update this section. https://github.com/ansible/awx/blob/devel/awx/ui_next/CONTRIBUTING.md#naming-files

jakemcdermott committed:
Fix relative imports

python tools/fixrelative.py awx/ui_next/src
npm --prefix=awx/ui_next run lint -- --fix
npm --prefix=awx/ui_next run prettier

jakemcdermott committed:
Rename .jsx -> .js

jakemcdermott committed:
Remove script to switch relative imports

jakemcdermott committed:
Relocate hooks to hooks directory

mkdir awx/ui_next/src/hooks
mv awx/ui_next/src/util/use* awx/ui_next/src/hooks/
find ./awx/ui_next/src -type f -exec sed -i -e 's/util\/use/hooks\/use/g' {} \;
find ./awx/ui_next/src/hooks -type f -exec sed -i -e 's/from \x27\.\//from \x27util\//g' {} \;
find ./awx/ui_next/src/hooks -type f -exec sed -i -e 's/from \x27util\/use/from \x27\.\/use/g' {} \;
npm --prefix=awx/ui_next run lint -- --fix
npm --prefix=awx/ui_next run prettier

jakemcdermott committed:
Regenerate translation string files

tiagodread reviewed (approved):
Completed an exploratory testing session on this branch and couldn't find anything broken so far all e2e tests looks good.

jakemcdermott commented:
regate

jakemcdermott commented:
regate
